src/test/incremental/hashes/for_loops.rs

```diff
@@ -94,7 +94,7 @@ pub fn change_iterable() {
 }
 
 #[cfg(not(cfail1))]
-#[rustc_clean(cfg="cfail2", except="HirBody, mir_built, optimized_mir")]
+#[rustc_clean(cfg="cfail2", except="HirBody, mir_built")]
```

Comment:
I tried to change this to `except="HirBody, mir_build, promoted_mir"` but I got this so I'm not sure this change is right:

```
error: `except` specified DepNodes that can not be affected for "ItemFn": "promoted_mir"
  --> /Users/wesley/code/rust/rust2/src/test/incremental/hashes/for_loops.rs:97:1
   |
LL | #[rustc_clean(cfg="cfail2", except="HirBody, mir_built, promoted_mir")]
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

thread '<unnamed>' panicked at 'assertion failed: `(left == right)`
  left: `LLVMing`,
 right: `Codegenning`', src/librustc_codegen_ssa/back/write.rs:1506:21
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace.
error: aborting due to previous error
```

Reply:
Are there any differences in the declaration of the optimized_mir query and the prmoted_mir query?

Reply:
None that I can see. The only thing I found was this https://github.com/rust-lang/rust/blob/ea52be482ab4945fda63cb65b6a198309a041e3c/src/librustc_incremental/persist/dirty_clean.rs#L73

but if I add `label_strs::promoted_mir` to that list, I get a ton of ICEs from the incremental tests.

Reply:
That looks very relevant. I think we just need to teach incremental about processing that query the same way as optimized_mir. What's the ICE about? I'd guess when loading the incremental cache it doesn't find the query

Reply:
Here's what I'm seeing:

```

failures:
    [incremental] incremental/hashes/call_expressions.rs
    [incremental] incremental/hashes/closure_expressions.rs
    [incremental] incremental/hashes/enum_constructors.rs
    [incremental] incremental/hashes/exported_vs_not.rs
    [incremental] incremental/hashes/for_loops.rs
    [incremental] incremental/hashes/function_interfaces.rs
    [incremental] incremental/hashes/if_expressions.rs
    [incremental] incremental/hashes/inherent_impls.rs
    [incremental] incremental/hashes/inline_asm.rs
    [incremental] incremental/hashes/let_expressions.rs
    [incremental] incremental/hashes/loop_expressions.rs
    [incremental] incremental/hashes/match_expressions.rs
    [incremental] incremental/hashes/panic_exprs.rs
    [incremental] incremental/hashes/struct_constructors.rs
    [incremental] incremental/hashes/unary_and_binary_exprs.rs
    [incremental] incremental/hashes/while_let_loops.rs
    [incremental] incremental/hashes/while_loops.rs
    [incremental] incremental/issue-42602.rs

test result: FAILED. 2 passed; 18 failed; 84 ignored; 0 measured; 0 filtered out

```

The failed tests all look like this one:

```
---- [incremental] incremental/issue-42602.rs stdout ----

error in revision `cfail2`: test compilation failed although it shouldn't!
status: exit code: 101
command: "/Users/wesley/code/rust/rust2/build/x86_64-apple-darwin/stage2/bin/rustc" "/Users/wesley/code/rust/rust2/src/test/incremental/issue-42602.rs" "-Zthreads=1" "--target=x86_64-apple-darwin" "--cfg" "cfail2" "-C" "incremental=/Users/wesley/code/rust/rust2/build/x86_64-apple-darwin/test/incremental/issue-42602/issue-42602.inc" "-Z" "incremental-verify-ich" "-Z" "incremental-queries" "--error-format" "json" "-Zui-testing" "-C" "prefer-dynamic" "--out-dir" "/Users/wesley/code/rust/rust2/build/x86_64-apple-darwin/test/incremental/issue-42602" "-Crpath" "-O" "-Cdebuginfo=0" "-Zunstable-options" "-Lnative=/Users/wesley/code/rust/rust2/build/x86_64-apple-darwin/native/rust-test-helpers" "-Zquery-dep-graph" "-L" "/Users/wesley/code/rust/rust2/build/x86_64-apple-darwin/test/incremental/issue-42602/auxiliary"
stdout:
------------------------------------------

------------------------------------------
stderr:
------------------------------------------
thread 'rustc' panicked at 'called `Option::unwrap()` on a `None` value', src/libcore/option.rs:378:21
stack backtrace:
   0: std::sys_common::backtrace::print
   1: std::panicking::default_hook::{{closure}}
   2: std::panicking::default_hook
   3: rustc::util::common::panic_hook
   4: std::panicking::rust_panic_with_hook
   5: std::panicking::continue_panic_fmt
   6: rust_begin_unwind
   7: core::panicking::panic_fmt
   8: core::panicking::panic
   9: rustc::dep_graph::graph::DepGraph::dep_node_index_of
  10: rustc_incremental::persist::dirty_clean::DirtyCleanVisitor::check_item
  11: rustc::hir::Crate::visit_all_item_likes
  12: rustc::dep_graph::graph::DepGraph::with_ignore
  13: rustc::dep_graph::graph::DepGraph::with_ignore
  14: rustc_incremental::persist::save::save_dep_graph
  15: rustc::util::common::time
  16: rustc_codegen_ssa::base::assert_and_save_dep_graph
  17: rustc_codegen_ssa::base::codegen_crate
  18: <rustc_codegen_llvm::LlvmCodegenBackend as rustc_codegen_utils::codegen_backend::CodegenBackend>::codegen_crate
  19: rustc::util::common::time
  20: rustc_interface::passes::start_codegen
  21: rustc::ty::context::tls::enter_global
  22: rustc_interface::passes::BoxedGlobalCtxt::access::{{closure}}
  23: rustc_interface::passes::create_global_ctxt::{{closure}}
  24: rustc_interface::passes::BoxedGlobalCtxt::enter
  25: rustc_interface::queries::Query<T>::compute
  26: rustc_interface::queries::<impl rustc_interface::interface::Compiler>::ongoing_codegen
  27: rustc_interface::interface::run_compiler_in_existing_thread_pool
  28: std::thread::local::LocalKey<T>::with
  29: scoped_tls::ScopedKey<T>::set
  30: syntax::with_globals
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
query stack during panic:
end of query stack

error: internal compiler error: unexpected panic

note: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/blob/master/CONTRIBUTING.md#bug-reports

note: rustc 1.39.0-dev running on x86_64-apple-darwin

note: compiler flags: -Z threads=1 -Z incremental-verify-ich -Z incremental-queries -Z ui-testing -Z unstable-options -Z query-dep-graph -C incremental -C prefer-dynamic -C rpath -C debuginfo=0


```

Reply:
So the ICE is coming from this line

https://github.com/rust-lang/rust/blob/e13fe7ff5e482b626714ac6ec41b8d44e7c22ed1/src/librustc/dep_graph/graph.rs#L426

Which is trying to load a `promoted_mir(def_id)` `DepNode` but it doesn't exist in the graph. I'm not sure yet why that would be because we do cache the `promoted_mir` query: 

https://github.com/rust-lang/rust/blob/73697266054308b3f920be30e97c0f7b3e409fb4/src/librustc/query/mod.rs#L133

As far as I can tell, the `promoted_mir` query is declared and configured in the same way as the `optimized_mir` query.

Reply:
It may not exist if its query was never executed, but idk how that should happen. I looked throughall the mentions of `optimized_mir`, but couldn't find anything either. @michaelwoerister any ideas?

The TLDR is that an incremental test suddenly doesn't show `optimized_mir` as dirty anymore, but I don't see how that can be. All that was done in this PR was to rip promoteds out of the `mir::Body`, but that should still cause MIR to be built (which it still does) and optimized (which it now doesn't).

Now maybe incremental tests don't codegen and thus never call optimized_mir, but then why was it being called before this PR? If it was called to get at the promoteds, that would explain the change, but I don'tknow what kind of code would get the promoteds but never the optimized mir body

Reply:
Do the optimized versions of MIR look the same in both revisions? We had cases where unoptimized MIR would look different but then the differences would get optimized away.

Reply:
Revision 1 doesn't have this function at all, as it is `cfg(not(cfail1))`. So the opimized `mir::Body` should be dirty in cfail2.

Reply:
The `cfg(cfail1)` version of the function is on line 88.

Reply:
:man_facepalming: ok, now everything makes sense.

The `&[0, 1, 2]` that is `&[0, 1, 3]` in `cfail1` is promoted. Since promoted's `mir::Body` is not part of the function's `mir::Body` anymore, `optimized_mir` is clean (it just references the promoted mir by id, which doesn't change across compilations).

So... the only question is now why the `promoted_mir` query doesn't show up here. The promoted mir is definitely changed.

Reply:
If I understand the code in `dirty_clean.rs` correctly, then `promoted_mir` is just not among the things that are checked for these kinds of tests.

Reply:
@michaelwoerister I tried adding it after `optimized_mir` locally on my computer here

https://github.com/rust-lang/rust/blob/ea52be482ab4945fda63cb65b6a198309a041e3c/src/librustc_incremental/persist/dirty_clean.rs#L73

but that caused the ICE's above.

Reply:
Explicitly listing `promoted_mir` in the `except` clause should work, I think.

Reply:
@wesleywiser Hm, yeah, the code seems to assume that there always is a dep-node for the given query/defid combination. That might not always be the case for `promoted_mir`?

Reply:
You could use `dep_graph.dep_node_exists()` to check if a [dep_graph.dep_node_index_of()](https://github.com/rust-lang/rust/blob/ea52be482ab4945fda63cb65b6a198309a041e3c/src/librustc_incremental/persist/dirty_clean.rs#L490) would fail. If it does not exist then `current_fingerprint` in the next line could be set to `None`. Would that make sense?

Reply:
> That might not always be the case for promoted_mir?

Yeah, I believe you are correct.

> You could use dep_graph.dep_node_exists() to check if a dep_graph.dep_node_index_of() would fail. If it does not exist then current_fingerprint in the next line could be set to None. Would that make sense?

That sounds reasonable to me. I'll give it a shot tonight.